src/sources/transient.rs

```diff
@@ -0,0 +1,672 @@
+///
+///    // Any other fields go here...
+/// }
+/// ```
+///
+/// To create the transient source, you can simply use the `Into`
+/// implementation:
+///
+/// ```none,actually-rust-but-see-https://github.com/rust-lang/rust/issues/63193
+/// let (sender, source) = channel();
+/// let mpsc_receiver: TransientSource<Channel> = source.into();
+/// ```
+///
+/// `TransientSource` implements [`EventSource`](crate::EventSource) and passes
+/// through `process_events()` calls, so in the parent's `process_events()`
+/// implementation you can just do this:
+///
+/// ```none,actually-rust-but-see-https://github.com/rust-lang/rust/issues/63193
+/// fn process_events<F>(
+///     &mut self,
+///     readiness: calloop::Readiness,
+///     token: calloop::Token,
+///     callback: F,
+/// ) -> Result<calloop::PostAction, Self::Error>
+/// where
+///     F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
+/// {
+///     let channel_return = self.mpsc_receiver.process_events(readiness, token, callback)?;
+///
+///     // Perform other logic here...
+///
+///     Ok(channel_return)
+/// }
+/// ```
+///
+/// Note that:
+///
+///   - You can call `process_events()` on the `TransientSource<Channel>` even
+///     if the channel has been unregistered and dropped. All that will happen
+///     is that you won't get any events from it.
+///
+///   - The [`PostAction`](crate::PostAction) returned from `process_events()`
+///     will only ever be `PostAction::Continue` or `PostAction::Reregister`.
+///     You will still need to combine this with the result of any other sources
+///     (transient or not).
+///
+/// Once you return `channel_return` from your `process_events()` method (and
+/// assuming it propagates all the way up to the event loop itself through any
+/// other event sources), the event loop might call `reregister()` on your
+/// source. All your source has to do is:
+///
+/// ```none,actually-rust-but-see-https://github.com/rust-lang/rust/issues/63193
+/// fn reregister(
+///     &mut self,
+///     poll: &mut calloop::Poll,
+///     token_factory: &mut calloop::TokenFactory,
+/// ) -> crate::Result<()> {
+///     self.mpsc_receiver.reregister(poll, token_factory)?;
+///
+///     // Other registration actions...
+///
+///     Ok(())
+/// }
+/// ```
+///
+/// The `TransientSource` will take care of updating the registration of the
+/// inner source, even if it actually needs to be unregistered or initially
+/// registered.
+#[derive(Debug)]
+pub enum TransientSource<T> {
+    /// The source should be kept in the loop.
+    Keep(Box<T>),
+    /// The source needs to be registered with the loop.
+    Register(Box<T>),
+    /// The source needs to be disabled but kept.
+    Disable(Box<T>),
+    /// The source needs to be removed from the loop.
+    Remove(Box<T>),
+    /// The source has been removed from the loop and dropped (this might also
+    /// be observed if there is a panic while changing states).
+    None,
+}
```

Comment:
Why the inner `Box`?

Reply:
This was a practical measure to reduce code size, because there's no implementation of `EventSource` for `Box<EventSource>` (or, say, some trait that `Box<EventSource>` would implement, I don't know if that's possible). I was using this in an application that had quite a few `TransientSource<SubprocSource>` where `SubprocSource` had a substantial on-stack size, so the box helped reduce code size and stack pressure in an embedded system. I'd be happy to remove it (it won't kill my application to go back, I was just trying lots of different things), but you'll probably see a PR from me in the near future implementing `EventSource` for `Box<EventSource>` or something like that.

Reply:
Oh, I see.

Well, an `impl<T> EventSource for Box<T> where T: EventSource { ... }` seems like a pretty legit thing to add to calloop, yeah. I'm not sure an impl for `Box<dyn EventSource>` would be possible though, because of the associated constants and all.

Reply:
Okay, I'll remove the inner boxing and submit a separate PR for the transitive-ness. The associated types shouldn't be a problem, see [Discord chat](https://discord.com/channels/442252698964721669/443150878111694848/949657615233847337).

Reply:
Ah yes indeed, looks like this kind of things work:

```rust
trait Foo {
    type Bar;
    
    fn method(&mut self, bar: Self::Bar);
}

impl<B> Foo for Box<dyn Foo<Bar=B>> {
    type Bar = B;
    
    fn method(&mut self, bar: B) {
        (&mut **self).method(bar)
    }
}
```